mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp

```diff
@@ -68,10 +65,21 @@ func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Loc
     }
 
     // Since the return value is guaranteed to be discarded, then let's change the return type
-    // to be only the quantum register.
+    // to be only the quantum register and the expval.
+    //
+    // We also need to return the expval to avoid dead code elimination downstream from
+    // removing the expval op in the body.
+    // TODO: we only support grad on expval op for now
     std::string fnName = callee.getName().str() + ".nodealloc";
     Type qregType = quantum::QuregType::get(rewriter.getContext());
-    FunctionType fnType = rewriter.getFunctionType(callee.getArgumentTypes(), qregType);
+
+    Type f64Type = rewriter.getF64Type();
+    SmallVector<Type> retTypes{qregType};
+    auto expvalOps = callee.getOps<quantum::ExpvalOp>();
+    std::for_each(expvalOps.begin(), expvalOps.end(),
```

Comment:
We should assert that our assumption is true, i.e. that there are no other measurement processes besides expval.